README.md

```diff
@@ -194,6 +194,32 @@ cargo clippy -- -A clippy::all -W clippy::useless_format -W clippy::...
 ```
 Note that if you've run clippy before, this may only take effect after you've modified a file or ran `cargo clean`.
 
+### Specifying the minimum supported Rust version
+
+Projects that intend to support old versions of Rust can disable lints pertaining to newer features by
+specifying the minimum supported Rust version (MSRV) in the clippy configuration file.
+
+```toml
+msrv = "1.30.0"
+```
+
+The MSRV can also be specified as an inner attribute, like below.
+
+```rust
+#![feature(custom_inner_attributes)]
+#![clippy::msrv = "1.30.0"]
+
+fn main() {
+  ...
+}
+```
+
+Tilde/Caret version requirements (like `^1.0` or `~1.2`) can be specified as well.
```

Comment:
IIRC [`semver::VersionReq` treats `1.1` as `^1.1` without `^`](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility), right? If so, these requirements are verbose and I recommend not to mention them. Also, I don't think it is necessary to mention `~`. MSRV is a specific version and is not a range. (I think it means strictly "it can be compiled with all stable versions after the specified version", but `~x.y` and some requirements are not the correct requirements to express this so I think should ideally be rejected.)

Reply:
> `~x.y` and some requirements are not the correct requirements to express this so I think should ideally be rejected.

I looked at how rustc's `cfg(version)` parses this, but it is worse: https://github.com/rust-lang/rust/issues/79436

I've commented about [an alternative parser implementation](https://github.com/rust-lang/rust/issues/79436#issuecomment-734057203), but I think that is probably sufficient for Clippy's use cases as well. If so, it may be preferable to use that instead of `VersionReq`. 
(I'm not the maintainer of Clippy, so I'll leave the decision to maintainers, but *I believe we shouldn't accept the extra requirement syntax that doesn't really make sense here*.)

Reply:
I don't see why the additional options to specify a version should be problematic. If you want to misuse those, that's on you. But I agree, that we should not mention that this is possible, otherwise it invites people to misuse it.

Reply:
@taiki-e yeah, the MSRV should definitely not be a range so tilde requirements don't make sense but, I don't think `VersionReq` treats `1.1` and `^1.1` the same way 

e.g. In the snippet below, switching between `1.30` and `^1.30` produces different results

```rust
use semver::{Version, VersionReq};

fn main() {
  const MANUAL_STRIP_MSRV: Version = Version {
      major: 1,
      minor: 45,
      patch: 0,
      pre: Vec::new(),
      build: Vec::new(),
  };
  match VersionReq::parse("1.30") {
    Ok(vReq) => {
      if !vReq.matches(&MANUAL_STRIP_MSRV) {
        println!("Will lint!")
      } else {
        println!("Won't lint!")
      }
    },
    Err(_) => println!("Error")
  }
}
```

Reply:
@Suyash458: Good catch! I didn't know that behavior. Given that [cargo treats them as the same requirements](https://doc.rust-lang.org/cargo/reference/resolver.html#semver-compatibility), I think it's preferable to treat the same as cargo.

@flip1995: I agree with it is a user issue, but the more things that can be accepted, the more complicated the test will be. I think it is preferable to accept only the minimum necessary things if possible.
In fact, @Suyash458 found that `"msrv = major.minor"` didn't seem to be working as expected. (Seems #6201 had tests for `msrv = "major.minor.patch"`, `msrv = "=major.minor.patch"`, `msrv = "^major.minor"`, etc., but not for `msrv = "major.minor"`.)

Reply:
> the more complicated the test will be. I think it is preferable to accept only the minimum necessary things if possible.

Well we use the semver crate, so this crate should have the tests, that the version matching works, so we can just assume, that it works. (And semver has indeed many unit tests)

Contrary to that: The semver documentation for [`VersionReq`](https://docs.rs/semver/0.11.0/semver/struct.VersionReq.html#examples-1) claims, that `1.30` is just an alias for `^1.30` and should therefore behave the same :sweat_smile:

Reply:
@taiki-e, @flip1995 `1.30` and `1.30.0` are treated the same way in `semver` v0.9 but this behavior is broken in v0.11